internal/conduit/registry/geth_block.go

```diff
@@ -18,81 +20,207 @@ const (
+		if oracle.cfg.StartHeight != nil {
+			log.Printf("StartHeight found to be: %d, using that value.", oracle.cfg.StartHeight)
+			return oracle.cfg.StartHeight
+		}
+
+		log.Printf("Starting Height is nil, using latest block as starting point.")
+		return nil
+	}
+
+	log.Printf("Currently processing height: %d", oracle.currHeight)
+	return oracle.currHeight
+}
+
+// fetchHeaderWithRetry ... retry for specified number of times.
+// Not an exponent backoff, but a simpler method which retries sooner
+func (oracle *GethBlockODef) fetchHeaderWithRetry(ctx context.Context, height *big.Int) *types.Header {
+	for i := 0; i <= oracle.cfg.NumOfRetries; i++ {
+		if i != 0 {
+			log.Printf("Header Retry number: %d", i)
+			time.Sleep(time.Duration(i) * time.Second)
+		}
+
+		header, err := oracle.clientInterface.HeaderByNumber(ctx, height)
+		if err != nil {
+			log.Printf("Header fetching error: %s", err.Error())
+		} else {
+			return header
+		}
+
+		if i == oracle.cfg.NumOfRetries {
+			log.Printf("All retries exhausted. Block at height %d will be skipped", height)
+			return nil
+		}
+	}
+	return nil
+}
+
+// fetchBlockWithRetry ... retry for specified number of times.
+// Not an exponent backoff, but a simpler method which retries sooner
+func (oracle *GethBlockODef) fetchBlockWithRetry(ctx context.Context, headerNumber *big.Int) *types.Block {
+	for i := 0; i <= oracle.cfg.NumOfRetries; i++ {
+		if i != 0 {
+			log.Printf("Block Retry number: %d", i)
+			time.Sleep(time.Duration(i) * time.Second)
+		}
+
+		block, err := oracle.clientInterface.BlockByNumber(ctx, headerNumber)
+		if err != nil {
+			log.Printf("Block fetching error: %s", err.Error())
+		} else {
+			return block
+		}
+
+		if i == oracle.cfg.NumOfRetries {
+			log.Printf("All retries exhausted. Block at height %d will be skipped", headerNumber)
+			return nil
+		}
+	}
 	return nil
 }
 
 // ReadRoutine ... Sequentially polls go-ethereum compatible execution
 // client using monotonic block height variable for block metadata
 // & writes block metadata to output listener components
 func (oracle *GethBlockODef) ReadRoutine(ctx context.Context, componentChan chan models.TransitData) error {
-	// NOTE - This poller logic is really bad and doesn't
-	//        currently compensate for a lot of edge cases, some of the obvious being:
-	// 1 - Client timeouts/failures; ie embed retry logic
-	// 2 - Only reads most recent headers and doesn't take a starting block # - DONE
-	// 3 - No optionality support for an ending block
+	// NOTE - Might need improvements in future as the project takes shape.
 
-	ticker := time.NewTicker(pollInterval * time.Millisecond)
+	if oracle.cfg.EndHeight != nil && oracle.cfg.StartHeight == nil {
+		return errors.New("cannot start with latest block height with end height configured")
+	}
+
+	if oracle.cfg.EndHeight.Cmp(oracle.cfg.StartHeight) < 0 {
+		return errors.New("start height cannot be more than the end height")
+	}
```

Comment:
Note that this should be changed in the future where these validations happen during Oracle construction so that pipelines are built out to then be invalidated

Reply:
Created #21